requires_renaming must not recurse into array subtypes in structs

tautschnig opened the pull request:
Just like pointers, array members may have subtypes of the same type as the
containing struct. The behaviour/check now matches what the actual renaming
function uses.

Also clean up the unnecessary ID_class, which does not actually exist.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
requires_renaming must not recurse into array subtypes in structs

Just like pointers, array members may have subtypes of the same type as the
containing struct. The behaviour/check now matches what the actual renaming
function uses.

Also clean up the unnecessary ID_class, which does not actually exist.

owen-mc-diffblue commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -622,16 +622,20 @@ static bool requires_renaming(const typet &type, const namespacet &ns)
   }
-  else if(
-    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_class)
+  else if(type.id() == ID_struct || type.id() == ID_union)
```
Is this change related to the rest of the PR?

tautschnig replied:
I guess I sneaked that one in - it just cleans up a dead branch that I had complained about back in https://github.com/diffblue/cbmc/pull/3633#discussion_r244408981. The agreed fix just never made it into this part of the code...

owen-mc-diffblue replied:
It would be better in a separate commit

tautschnig replied:
Fair enough - that's now done.

tautschnig committed:
Remove impossible ID_class case

An input-language "class" is internally a struct (with an ID_C_class)
annotation. The other parts of the renaming code have previously already had
this dead case removed.

kroening commented:
So what would happen to
```
struct S
{
  struct
  {
    int something[x];
  } whatnot[2];
};
```
I think we need to keep a set of struct tags that we have seen already around.

kroening commented:
Or I should finish the code in branch remove-flexible-array-member

tautschnig commented:
Yes, renaming is part broken for such examples. Here is a full example, with current develop (not even this branch!):
```
$ cat ex.c
int main()
{
  int x = 1;

  struct S
  {
    struct
    {
      int something[x];
    } whatnot[2];
  } s;

  s.whatnot[0].something[0] = 42;
}
$ cbmc ex.c
CBMC version 5.11 (cbmc-5.11-1583-g51f5796) 64-bit x86_64 linux
Parsing ex.c
[...]
Starting Bounded Model Checking
--- begin invariant violation report ---
Invariant check failed
File: ../src/util/simplify_expr_struct.cpp:51 function: simplify_member
Condition: base_type_eq(op2.type(), expr.type(), ns)
Reason: member expression type must match component type
Backtrace:
[...]
```

tautschnig committed:
Failing test to show that requires_renaming recurses

requires_renaming must not recurse into array subtypes in structs.

tautschnig commented:
This PR will be redundant if we decide to go with #4286 (which now also includes the test added here).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 95c2c27a45088665b27ac6dd0d2d221a82cc6bc9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105812595
